### PR TITLE
[rfc] GL api wrapper

### PR DESCRIPTION
**Pull request by hjanetzek**
This should allow to cleanly use the evas_glapi on Tizen.

We may also consider to directly use the platform GL headers again. At least for me compile times are not an issue anymore :)

### CODE REVIEW

File: core/src/gl.h [hunk 1]
```diff
@@ -270,156 +248,121 @@ typedef char            GLchar;
...
-    GL_APICALL void GL_APIENTRY glStencilMask(GLuint mask);
-    GL_APICALL void GL_APIENTRY glStencilOp(GLenum fail, GLenum zfail, GLenum zpass);
-    GL_APICALL void GL_APIENTRY glClearStencil(GLint s);
-    GL_APICALL void GL_APIENTRY glColorMask(GLboolean red, GLboolean green, GLboolean blue, GLboolean alpha);
-    GL_APICALL void GL_APIENTRY glCullFace(GLenum mode);
-    GL_APICALL void GL_APIENTRY glFrontFace(GLenum mode);
-    GL_APICALL void GL_APIENTRY glClearColor(GLclampf red, GLclampf green, GLclampf blue, GLclampf alpha);
-    GL_APICALL void GL_APIENTRY glGetIntegerv(GLenum pname, GLint *params );
+namespace Tangram {
+struct GL {
```
Comment (tallytalwar) [inline core/src/gl.h]: @hjanetzek and I were discussing yesterday about `struct` vs `namespace` for `GL` here. The only thing `struct` brings here is future proofing for if we want to add any member variable (maybe some debugging).

**Comment by tallytalwar** [comment]
Tested this on `android` and `osx`. I can not get `ios` to build on my machine, @blair1618 can you test this out, please?

Thanks

**Comment by tallytalwar** [comment]
I am good with this. :+1:

**Comment by blair1618** [comment]
Checkin it out

**Comment by blair1618** [comment]
No issues building and running on iOS. More thoughts on this in a bit.

File: CMakeLists.txt [hunk 2]
```diff
@@ -4,6 +4,7 @@ project(tangram)
 
 # Options
 option(USE_SYSTEM_FONT_LIBS "Use system libraries Freetype, ICU and Harfbuzz via pkgconfig" OFF)
+option(USE_SYSTEM_GLFW_LIBS "Use system libraries for GLFW3 via pkgconfig" OFF)
```
Comment (blair1618) [inline CMakeLists.txt]: Why is this change in this PR? It's not a bad change, just seems unrelated.

File: core/common/platform_gl.cpp [hunk 3]
```diff
@@ -0,0 +1,287 @@
+#include "platform_gl.h"
+
+#include "gl.h"
+#include "gl/error.h"
+
+namespace Tangram {
+
+GLenum GL::getError() {
+    auto result = glGetError();
+    GL_CHECK();
```
Comment (blair1618) [inline core/common/platform_gl.cpp]: I think this will cause unbounded recursion into `GL::getError`

File: core/common/platform_gl.h [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+#ifndef GL_GLEXT_PROTOTYPES
+#define GL_GLEXT_PROTOTYPES 1
+#endif
+
+#include <GLES2/gl2.h>
+#include <GLES2/gl2ext.h>
+typedef void (GL_APIENTRYP PFNGLBINDVERTEXARRAYOESPROC) (GLuint array);
+typedef void (GL_APIENTRYP PFNGLDELETEVERTEXARRAYSOESPROC) (GLsizei n, const GLuint *arrays);
+typedef void (GL_APIENTRYP PFNGLGENVERTEXARRAYSOESPROC) (GLsizei n, GLuint *arrays);
+typedef GLboolean (GL_APIENTRYP PFNGLISVERTEXARRAYOESPROC) (GLuint array);
```
Comment (blair1618) [inline core/common/platform_gl.h]: I believe these types are already declared in `gl2ext.h`

**Comment by blair1618** [comment]
This is cool. I like that GL error checking is now consolidated to one file and that macros and typedefs are hidden from GL function call sites.

What I'd like to move towards eventually is using the official Khronos GL ES 2 headers and a dynamic function loader like [GLAD](https://github.com/Dav1dde/glad). For the platforms that don't ship with ES 2 we can load ANGLE libraries for a consistent interface. 

In the meantime this looks like a good step that certainly simplifies the Tizen GL interface.

File: CMakeLists.txt (hunk 2 above)
Comment (hjanetzek) [inline CMakeLists.txt]: I can move it to a separate PR if you prefer.

File: core/common/platform_gl.cpp (hunk 3 above)
Comment (hjanetzek) [inline core/common/platform_gl.cpp]: Good catch!

Comment (tallytalwar) [inline core/common/platform_gl.cpp]: Good eye!!

**Comment by blair1618** [comment]
LGTM
